src/Symfony/Component/Security/Core/Role/RoleHierarchy.php

```diff
@@ -76,6 +74,49 @@ protected function buildRoleMap()
             }
 
             $this->map[$main] = array_unique($this->map[$main]);
+
+            if (str_contains($main, '*')) {
+                $this->rolePlaceholdersPatterns[$main] = sprintf('/%s/', strtr($main, ['*' => '[^\*]+']));
+            }
         }
     }
+
+    private function resolveReachableRoleNames(array $roles, array &$visitedPlaceholders = []): array
+    {
+        $reachableRoles = $roles;
+
+        foreach ($roles as $role) {
+            if (!isset($this->map[$role])) {
+                continue;
+            }
+
+            foreach ($this->map[$role] as $r) {
+                $reachableRoles[] = $r;
+            }
+        }
+
+        $placeholderRoles = array_diff($this->getMatchingPlaceholders($reachableRoles), $visitedPlaceholders);
+        if (!empty($placeholderRoles)) {
+            array_push($visitedPlaceholders, ...$placeholderRoles);
+            $resolvedPlaceholderRoles = $this->resolveReachableRoleNames($placeholderRoles, $visitedPlaceholders);
+            foreach (array_diff($resolvedPlaceholderRoles, $placeholderRoles) as $r) {
+                $reachableRoles[] = $r;
+            }
+        }
+
+        return array_values(array_unique($reachableRoles));
```

Comment:
Do we really need to do the uniqueness filtering in the method that gets called recursively ? couldn't we do it only once at the end of the recursion ?

Reply:
No you're right, I'll fix that in the next commit, thanks

Reply:
I force pushed this one to let the next commit about regex cleaner